lib/workflows/send_errored_vm_email.py

```diff
@@ -0,0 +1,180 @@
+
+from structs.email.email_params import EmailParams
+from structs.email.email_template_details import EmailTemplateDetails
+from structs.email.smtp_account import SMTPAccount
+
+
+def find_servers_with_errored_vms(
+    cloud_account: str, time_variable:int, from_projects: Optional[List[str]] = None
+):
+    """
+    :param cloud_account: string represents cloud account to use
+    :param from_projects: A list of project identifiers to limit search in
+    """
+
+
+    server_query = ServerQuery()
+    if time_variable > 0:
+        server_query.where(
+            QueryPresetsDateTime.OLDER_THAN, ServerProperties.SERVER_LAST_UPDATED_DATE, days=time_variable
+        )
+    server_query.where(
+        QueryPresetsGeneric.ANY_IN, ServerProperties.SERVER_STATUS, values=["ERROR"]
+    )
+    server_query.run(
+        cloud_account,
+        as_admin=True,
+        from_projects=from_projects if from_projects else None,
+        all_projects=not from_projects,
+    )
+
+    server_query.select("id", "name", "addresses")
+
+    if not server_query.to_props():
+        raise RuntimeError(
+            f"No servers found in [ERROR] state on projects "
+            f"{','.join(from_projects) if from_projects else '[all projects]'}"
+        )
+
+    server_query.append_from("PROJECT_QUERY", cloud_account, "name")
+    server_query.group_by("user_id")
+
+    return server_query
+
+
+def print_email_params(
+    email_addr: str, user_name: str, as_html: bool, error_table: str
+):
+    """
+    Print email params instead of sending the email
+    :param email_addr: email address to send to
+    :param user_name: name of user in openstack
+    :param as_html: A boolean which if selected will send an email, otherwise prints email details only
+    :param error_table: a table representing info found in openstack
+    about VMs in error state
+    """
+    print(
+        f"Send Email To: {email_addr}\n"
+        f"email_templates errored-email: username {user_name}\n"
+        f"send as html: {as_html}\n"
+        f"error table: {error_table}\n"
+    )
+
+
+def build_email_params(user_name: str, error_table: str, **email_kwargs):
+    """
+    build email params dataclass which will be used to configure how to send the email
+    :param user_name: name of user in openstack
+    :param error_table: a table representing info found in openstack about VMs
+        that are in errored state
+    :param email_kwargs: a set of email kwargs to pass to EmailParams
+    """
+    body = EmailTemplateDetails(
+        template_name="errored_vm",
+        template_params={
+            "username": user_name,
+            "error_table": error_table,
+        },
+    )
+
+    footer = EmailTemplateDetails(template_name="footer", template_params={})
+
+    return EmailParams(email_templates=[body, footer], **email_kwargs)
+
+
+def find_user_info(user_id, cloud_account, override_email_address):
```

Comment:
Missing return type here, though I'm not sure what `res["user_name"][0], res["user_email"][0]` is.

Reply:
Relevant return type added.